nbbuild/antsrc/org/netbeans/nbbuild/extlibs/CreateNoticeFile.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.nbbuild.extlibs;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileFilter;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.Set;
+import java.util.StringTokenizer;
+import java.util.TreeSet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.tools.ant.BuildException;
+import org.apache.tools.ant.Task;
+import org.apache.tools.ant.types.FileSet;
+
+/**
+ * Creates an Apache compliant NOTICE file.
+ *
+ * Apache NOTICE files must always start with a preamble. This is specified in
+ * the "preamble" property, usually set to "${nb_all}/nbbuild/notice-stub.txt".
+ *
+ * The NOTICE file must include the NOTICE files of all external binaries that
+ * have one. As a consequence the NOTICE file is different for a platform build
+ * than for an ide build. When "binaryBuild" is set to true (the default) this
+ * task will scan all the appropriate modules in the build and automatically
+ * include the "[module]external/*notice.txt" and "[module]/*notice.txt" files,
+ * if they exist.
+ *
+ * If "binaryBuild" is not set to true then only the preamble and the
+ * "[module]/*notice.txt" files will be included in the final NOTICE file.
+ *
+ */
+public class CreateNoticeFile extends Task {
```

Comment:
Please note that CreateLicenseSummary is already constructing a NOTICE file from notice-stub and the -notice.txt in used binary dependencies. The rest (i.e. source-based -notice.txt files) are added to the NOTICE file in the build script. Moving that to Java is OK (and maybe desirable), but I think it would be better to merge that into CreateLicenseSummary, or at least have a common utilities/code, as for binary builds, CreateLicenseSymmary already does a lot of checks, which would need to be duplicated here.

Reply:
The ${nbbuild.dir}/notice-temp being generated by CreateLicenseSummary has many duplicated entries (Apache Ant appears 24 times, 'mecab' appears 4 times). We'll have to look at that. Also we're computing the CRC sums of jars, zips & xml's under nbbuild/build, not sure we want to do that, though.

Maybe we want to compute the CRC sums of jars under ${allmodules}/external/*.jar only? I'll try to look at CreateLicenseSummary and see what I can modify. Some requirements would be appreciated ;-)

Reply:
I see now why scanning all binaries is required. Quite clever!